File: tester/Fx_Controller.js
"use strict";

const Fx_Controller_Clip_Count = 6;
var Fx_Controller_PreviousMillis = 0;
var Fx_Controller_Now = 0;

function Fx_Controller(NUMBER_OF_PIXELS) {
  this.NUMBER_OF_PIXELS = NUMBER_OF_PIXELS;
  this.foregroundPixels = new Array(NUMBER_OF_PIXELS);
  this.backgroundPixels = new Array(NUMBER_OF_PIXELS);
  this.reset();
}

Fx_Controller.prototype.reset = function() {
  this.clips = new Array(Fx_Controller_Clip_Count);
  for (let i=0; i<Fx_Controller_Clip_Count; i++) {
    this.clips[i] = null;
  }
  this.lastClipIndex = -1;
  this.now = null;
  for (let i=0; i<NUM_PIXELS; i++) {
    this.foregroundPixels[i] = {r:0,g:0,b:0};
    this.backgroundPixels[i] = {r:0,g:0,b:0};
  }
  console.log("Fx_Controller reset, lastClipIndex: ", this.lastClipIndex, this.clips);
};

Fx_Controller.prototype.start = function() {
  Fx_Controller_Now = millis();
  Fx_Controller_PreviousMillis = millis();
};

// onFrame / perFrame / updateForFrame??
Fx_Controller.prototype.updateTimeline = function() {
  let now = millis();
  if (Fx_Controller_PreviousMillis > now) {
    // overflow happened
    Fx_Controller_PreviousMillis = Fx_Controller_Now = now;
  }
  Fx_Controller_PreviousMillis = Fx_Controller_Now;
  Fx_Controller_Now = now;

  // go over the clips list
  // prune out elapsed clips, re-start repeating clips
  for (let i=0; i<=this.lastClipIndex; i++) {
    let clip = this.clips[i];
    console.log("updating clip, is isForeground", clip.isForeground);
    let fillForeground = false;
    let fillBackground = false;
    if (!clip) {
      console.log("updateTimeline: falsey clip at index: ", i, this.clips);
    }
    let elapsed = now - clip.startTime;
    if (elapsed >= clip.duration) {
      clip.repeat--;
      clip.startTime = now;
    }
    if (clip.repeat < 0) {
      console.log("updateTimeline: clip with animateFn " + clip.animateFn.name + " is out of repeats, removing it from index: " + i);
      this.removeClipAtIndex(i);
      i--;
      // reset the buffers to avoid lingering values from removed animation
      if (clip.isForeground) {
        this._fillForeground(0);
      } else {
        this._fillBackground(0);
      }
      continue;
    }
      // update the tracks with their new values
    let progress = 0;
    if (elapsed > 0) {
      progress = elapsed / clip.duration;
    }
    if (clip.isForeground) {
      clip.animateFn(progress,
                    this.foregroundPixels,
                    this.NUMBER_OF_PIXELS,
                    clip.params);
    } else {
      clip.animateFn(progress,
                    this.backgroundPixels,
                    this.NUMBER_OF_PIXELS,
                    clip.params);
    }
  }
};

Fx_Controller.prototype.addClip = function(clip) {
  let insertIndex = 1 + this.lastClipIndex;
  if (insertIndex >= Fx_Controller_Clip_Count) {
    // if we filled the array of clips, just replace the last one
    insertIndex = this.lastClipIndex;
  }
  this.clips[insertIndex] = clip;
  this.lastClipIndex = insertIndex;
  console.log("addClip: ", this.clips[insertIndex]);
}

Fx_Controller.prototype.removeClipAtIndex = function(removeIdx) {
  removeIdx = clamp(removeIdx, 0, Fx_Controller_Clip_Count - 1);
  let lastIndex = Fx_Controller_Clip_Count - 1;
  this.clips[removeIdx] = null;
  for (let i = removeIdx+1; i <= this.lastClipIndex; i++) {
    this.clips[i -1] = this.clips[i];
  }
  this.lastClipIndex--;
  console.log("removeClipAtIndex result: ", this.lastClipIndex, this.clips);
}

Fx_Controller.prototype._fillForeground = function(value = 0) {
  for(let i=0; i<NUM_PIXELS; i++) {
    let pixel = fxController.foregroundPixels[i];
    pixel.r = pixel.g = pixel.b = value;
  }
};

Fx_Controller.prototype._fillBackground = function(value = 0) {
  for(let i=0; i<NUM_PIXELS; i++) {
    let pixel = fxController.backgroundPixels[i];
    pixel.r = pixel.g = pixel.b = value;
  }
};


function Fx_Controller_Clip(animation, startTime, duration, repeat, params) {
  /*
    a clip is an animation function, with a start time and duration
    the clip tracks the number of times it will repeat
    and passes a set of params into that animation on each call
  */
  this.animateFn = animation.animateFn;
  console.log("Fx_Controller_Clip, got params: ", params);
  // lift up the animation default params to be instance state for this sequence
  this.params = Object.assign({}, animation.defaulParams, params);
  this.duration = duration;
  this.startTime = startTime;
  this.repeat = repeat || 0;
}

File: led-arch/lib/Fx_Controller.cpp
#include "Fx_Types.h"
#include "Fx_Helpers.h"
#include "Fx_Controller.h"
#include "Fx_Animations.h"

Fx_Controller fxController;
// allocate memory for each clip in the array
Fx_Controller_Clip clipSlots[MAX_FX_CLIPS];

void Fx_Controller_Reset() {
  fxController.lastClipIndex = -1;
  fxController.previousMillis = fxController.now = millis();
  RGBColor pixel;
  for (unsigned char i=0; i<fxController.clipCount; i++) {
    pixel = fxController.foregroundPixels[i];
    pixel.r = pixel.g = pixel.b = 0;
    pixel = fxController.backgroundPixels[i];
    pixel.r = pixel.g = pixel.b = 0;
  }
}

void Fx_Controller_Init() {
  fxController.pixelCount = NUM_PIXELS;
  fxController.clipCount = MAX_FX_CLIPS;

  // populate.clips with pointers to (static) memory allocated to clips
  for(unsigned char i=0; i<MAX_FX_CLIPS; i++) {
    fxController.clips[i] = &clipSlots[i];
  }

  Fx_Controller_Reset();
}

void Fx_Controller_updateTimeline(unsigned long now) {
  fxController.now = now;
  if ((fxController.previousMillis > fxController.now) ||
      fxController.previousMillis == 0) {
    // first update, or maybe overflow happened
    fxController.previousMillis = fxController.now;
  }

  // go over the clips list
  // prune out elapsed clips, re-start repeating clips
  for (unsigned char i=0; i<=fxController.lastClipIndex; i++) {
    Fx_Controller_Clip *clip = fxController.clips[i];
    unsigned long elapsed = fxController.now - clip->startTime;

    //debugPrint("updateTimeline clip at index %d: elapsed: %li, duration: %li\n", i, elapsed, clip->duration);

    Assert(clip->startTime > 0);
    Assert(clip->repeat >= 0);

    if (elapsed >= (unsigned long)clip->duration) {
      debugPrint("updateTimeline: animation complete, elapsed: %ld, repeat: %ld\n", elapsed, clip->repeat);
      clip->startTime = fxController.now;
      clip->repeat--;
      elapsed = elapsed - clip->duration;
      debugPrint("updateTimeline: decremented repeat %ld\n", clip->repeat);
    } else {
      //debugPrint("updateTimeline: didnt exceed duration\n");
    }
    if (clip->repeat < 0) {
      debugPrint("updateTimeline: clip is out of repeats, removing it from index: %d\n", i);
      Fx_Controller_RemoveClipAtIndex(i);
      i--;
      // reset the buffers to avoid lingering values from removed animation
      if (clip->isForeground) {
        Fx_Controller_FillForeground(0);
      } else {
        Fx_Controller_FillBackground(0);
      }
      continue;
    } else {
      //debugPrint("updateTimeline: can still repeat: %ld\n", clip->repeat);
    }
    // update the tracks with their new values
    float progress = 0.0;
    if (clip->duration > 0) {
      if (elapsed > 0) {
        progress = (float)elapsed / (float)clip->duration;
        //debugPrint("elapsed is > 0.0: %ld, progress: %.2f\n", elapsed, progress);
      }
    } else {
      // avoid divide by zero with duration: 0
      progress = 1.0;
    }
    progress = clampFloat(progress, 0.0, 1.0);
    Assert(progress <= 1 && progress >= 0);

    //debugPrint("elapsed: %ld, duration: %d, progress: %.2f\n", elapsed, clip->duration, progress);
    if (clip->isForeground) {
      // debugPrint("calling foreground animateFn with progress: %.2f\n", progress);
      (*clip->animateFn)(progress,
                         fxController.foregroundPixels,
                         fxController.pixelCount,
                         &clip->params);
    } else {
      // debugPrint("calling background animateFn with progress: %.2f\n", progress);
      (*clip->animateFn)(progress,
                         fxController.backgroundPixels,
                         fxController.pixelCount,
                         &clip->params);
    }
  }
  fxController.previousMillis = fxController.now;
}

void Fx_Controller_AddClip(AnimateFnPointer animateFn,
                     bool isForeground,
                     long duration,
                     long repeat,
                     Fx_AnimationParams params) {

  debugPrint("Fx_Controller_AddClip, lastClipIndex: %d\n", fxController.lastClipIndex);
  unsigned char insertIndex = 1 + fxController.lastClipIndex;
  if (insertIndex >= fxController.clipCount) {
    // if we filled the array of clips, just replace the last one
    insertIndex = fxController.lastClipIndex;
  }
  Assert(fxController.now > 0);

  // we have a fixed number of slots for clips and just re-use (overwrite) them
  Fx_Controller_Clip *clip = fxController.clips[insertIndex];
  clip->animateFn = animateFn;
  clip->isForeground = isForeground;
  clip->duration = duration;
  clip->repeat = repeat;
  clip->params = params;
  clip->startTime = fxController.now;

  debugPrint("Adding clip at index: %d, startTime: %ld\n", insertIndex, clip->startTime);
  fxController.lastClipIndex = insertIndex;
  debugPrint("Fx_Controller_AddClip at lastClipIndex result: %d\n", fxController.lastClipIndex);
  debugPrint("Fx_Controller_AddClip, startTime: %ld, duration: %ld, repeat: %ld\n",
      fxController.clips[insertIndex]->startTime,
      fxController.clips[insertIndex]->duration,
      fxController.clips[insertIndex]->repeat);
}

void Fx_Controller_RemoveClipAtIndex(unsigned char removeIdx) {
  // clips is a fixed-length array,
  // so memory is not freed when removing a clip, we just move the lastClipIndex
  // to allow it to be reused
  removeIdx = clamp(removeIdx, 0, fxController.clipCount - 1);
  for (unsigned char i = removeIdx+1; i <= fxController.lastClipIndex; i++) {
    fxController.clips[i -1] = fxController.clips[i];
  }
  fxController.lastClipIndex--;
  debugPrint("removeClipAtIndex remove from: %d, lastClipIndex: %d\n", removeIdx, fxController.lastClipIndex);
}

void Fx_Controller_FillForeground(unsigned char value) {
  for(unsigned short i=0; i<fxController.pixelCount; i++) {
    RGBColor pixel = fxController.foregroundPixels[i];
    pixel.r = pixel.g = pixel.b = value;
  }
}

void Fx_Controller_FillBackground(unsigned char value) {
  for(unsigned short i=0; i<fxController.pixelCount; i++) {
    RGBColor pixel = fxController.backgroundPixels[i];
    pixel.r = pixel.g = pixel.b = value;
  }
}

